ttnn/cpp/ttnn/operations/eltwise/binary_ng/device/binary_ng_device_operation.cpp

```diff
@@ -0,0 +1,235 @@
+                output_shape[i + larger_rank] = dim_a + dim_b - 1;
+            }
+        }
+        return ttnn::SimpleShape(output_shape);
+    };
+
+    auto output_shape = compute_broadcasted_output(input_shape_a, input_shape_b);
+    return TensorSpec(output_shape, TensorLayout(attributes.dtype, PageConfig(Layout::TILE), attributes.memory_config));
+}
+
+BinaryNgDeviceOperation::program_factory_t BinaryNgDeviceOperation::select_program_factory(
+    const operation_attributes_t&, const tensor_args_t&) {
+    return ProgramFactory{};
+}
+
+BinaryNgDeviceOperation::tensor_return_value_t BinaryNgDeviceOperation::create_output_tensors(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    const auto& output_tensor = tensor_args.output_tensor;
+    if (output_tensor.has_value()) {
+        return output_tensor.value();
+    }
+
+    return create_device_tensor(
+        compute_output_specs(operation_attributes, tensor_args), tensor_args.input_tensor_a.device());
+}
+
+tt::stl::hash::hash_t BinaryNgDeviceOperation::compute_program_hash(
+    const operation_attributes_t& attributes, const tensor_args_t& tensor_args) {
+    const auto& input_tensor_a = tensor_args.input_tensor_a;
+    const auto& input_tensor_b = tensor_args.input_tensor_b;
+
+    TT_ASSERT(
+        std::holds_alternative<DeviceStorage>(input_tensor_a.get_storage()),
+        "Unexpected type {}",
+        tt::stl::get_active_type_name_in_variant(input_tensor_a.get_storage()));
+
+    if (input_tensor_b.has_value()) {
+        TT_ASSERT(
+            std::holds_alternative<DeviceStorage>(input_tensor_b->get_storage()),
+            "Unexpected type {}",
+            tt::stl::get_active_type_name_in_variant(input_tensor_b->get_storage()));
+
+        return operation::hash_operation<BinaryNgDeviceOperation>(
+            attributes,
+            input_tensor_a.dtype(),
+            std::get<DeviceStorage>(input_tensor_a.storage()).memory_config(),
+            input_tensor_b->dtype(),
+            std::get<DeviceStorage>(input_tensor_b->storage()).memory_config());
+    }
+
+    return operation::hash_operation<BinaryNgDeviceOperation>(
+        attributes, input_tensor_a.dtype(), std::get<DeviceStorage>(input_tensor_a.storage()).memory_config());
+}
+
+std::tuple<BinaryNgDeviceOperation::operation_attributes_t, BinaryNgDeviceOperation::tensor_args_t>
+BinaryNgDeviceOperation::invoke(
+    const Tensor& input_tensor_a_arg,
+    const Tensor& input_tensor_b_arg,
+    BinaryOpType binary_op_type,
+    const std::optional<const DataType>& output_dtype,
+    const std::optional<MemoryConfig>& memory_config,
+    std::optional<Tensor> optional_output_tensor) {
+    if (output_dtype.has_value() && optional_output_tensor.has_value()) {
+        TT_FATAL(
+            output_dtype.value() == optional_output_tensor.value().get_dtype(),
+            "If both output dtype and output tensor provided dtype should match");
+    }
```

Comment:
Ah i see validation for output tensor here. Would it make more sense for this to be inside the op?